v2/blockstore/readwrite.go

```diff
@@ -260,7 +261,40 @@ func (b *ReadWrite) Finalize() error {
 }
 
 func (b *ReadWrite) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
-	return b.ronly.AllKeysChan(ctx)
+	if b.opts.BlockstoreUseWholeCIDs {
+		// slow path
+		return b.ronly.AllKeysChan(ctx)
+	}
+
+	if ctx.Err() != nil {
+		return nil, ctx.Err()
+	}
+
+	b.ronly.mu.Lock()
+	defer b.ronly.mu.Unlock()
+
+	if b.ronly.closed {
+		return nil, errClosed
+	}
+
+	out := make(chan cid.Cid)
+
+	go func() {
+		defer close(out)
+		err := b.idx.ForEach(func(mh multihash.Multihash, _ uint64) error {
+			select {
+			case out <- cid.NewCidV1(cid.Raw, mh):
```

Comment:
I think this probably breaks contract. It'll be fine if you're using the keys to talk back to the blockstore, but the contract doesn't specify that this is the only use of `AllKeysChan` I think, so this probably isn't OK.

But I think there is a way you can still achieve what you want. `idx` here is always `InsertionIndex` which is an internal construct that we don't leak, so you could add a `ForEachCid` to internal/store/insertionindex.go and do this loop but return the actual CIDs. Then you don't even need the `BlockstoreUseWholeCIDs` check. (I toyed with exactly this while working my way toward the eventual v2/storage implementation where it turned out I didn't need it).

We could potentially do something roughly the same in ReadOnly when we do an index generation and return an index (maybe InsertionIndex, or something else) that has a `ForEachCids` and then do a feature/type check on the index to see if we don't have to iterate because it ends up a bit strange that a ReadWrite has a fast path but ReadOnly doesn't. I won't burden you with needing to do that here if you don't want but a TODO mentioning the possibility of making ReadOnly have a fast path might be good.

Reply:
> I think this probably breaks contract.

Could you express that problem in the form of a go test? I don't see what I'm breaking here. Is that it's not necessarily the same codec?

Reply:
i think one issue is that this always says the keys are of raw multicodec - e.g. it drops the codec part of the CID.

this means that other ipld utilities that may expect to e.g. attempt to parse dagPB to understand dag structure or otherwise care about the codec could end up with different behavior from before.

Reply:
@willscott thanks for the explanation. FYI though, it's not in the latest code anymore.